core/drivers/amd/mailbox_driver.c

```diff
@@ -0,0 +1,216 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2024-2025, Advanced Micro Devices, Inc. All rights reserved.
+ *
+ */
+#include <drivers/amd/mailbox_driver.h>
+#include <initcall.h>
+#include <kernel/delay.h>
+#include <kernel/panic.h>
+#include <mailbox_private.h>
+#include <malloc.h>
+#include <mm/core_mmu.h>
+#include <platform_config.h>
+#include <string.h>
+#include <tee/cache.h>
+
+static struct ipi_info_t ipi_info;
+
+static TEE_Result mailbox_call(enum mailbox_api ipi_api, uint32_t blocking)
+{
+	struct thread_smc_args args = {
+		.a0 = PLAT_SMC_SIP_IPI | ipi_api,
+		.a1 = reg_pair_to_64(0, ipi_info.local),
+		.a2 = reg_pair_to_64(0, ipi_info.remote),
+		.a3 = reg_pair_to_64(0, blocking),
+	};
+
+	thread_smccc(&args);
+
+	if (args.a0) {
+		EMSG("Mailbox Call returned with error 0x%lx", args.a0);
+		return TEE_ERROR_GENERIC;
+	}
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result mailbox_open(uint32_t remote_id, size_t payload_size)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	paddr_t req_addr = 0, rsp_addr = 0;
+
+	ipi_info.remote = remote_id;
+
+	req_addr = (ipi_info.buf + (remote_id * IPI_OFFSET_MULTIPLIER) +
+		    IPI_REQ_OFFSET);
+
+	rsp_addr = (ipi_info.buf + (remote_id * IPI_OFFSET_MULTIPLIER) +
+		    IPI_RESP_OFFSET);
```

Comment:
Nitpicking: could remove inner and outer parentheses in both instructions above.